zdb extensions for zstd

allanjude opened the pull request:
Add an extension to zdb to be able to decode the zstd compression header

Moves the zstd_header definitions to the .h file so it can be reused in zdb

Also fixes up the comments around `zstd_enum_to_cookie()`

allanjude committed:
Move the zstd compression header to the .h file so it can be reused

Sponsored-By: The FreeBSD Foundation
Signed-off-by: Allan Jude <allanjude@freebsd.org>

allanjude committed:
Replace the term cookie

In the original version of this code, the 'cookie' was the compression
header, that is now zstd_header_t. Over the course of its evolution
the variable 'cookie' has become the signed integer level passed to
ZSTD for compression, so rename it to what it actually represents.

While here, fix the comment around zstd_enum_to_cookie(),
It incorrectly stated:
"Convert ZSTD level to ZFS internal, stored level"

when it does the opposite, converting the internal ZFS enum to
the signed integer used by ZSTD.

Sponsored-By: The FreeBSD Foundation
Signed-off-by: Allan Jude <allanjude@freebsd.org>

allanjude committed:
zdb: add a new flag (-Z) to inspect ZSTD compression header

zdb can now decode the ZSTD compression header and inspect the size,
version, and compression level saved in that header.

This will be used to create a test that will ensure this information
is saved and restored correctly.

Example usage:
```
obj=2 dataset=zof/zstd/10 path=/allanbsd.txt type=19 bonustype=44

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         2    1   128K  8.50K      0     512  8.50K  100.00  ZFS plain file (K=inherit) (Z=inherit)
                                               168   bonus  System attributes
        dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED
        dnode maxblkid: 0
        uid     0
        gid     0
        atime   Sat Jun 13 16:03:26 2020
        mtime   Sat Jun 13 16:03:26 2020
        ctime   Sat Jun 13 16:03:26 2020
        crtime  Sat Jun 13 16:03:26 2020
        gen     3046
        mode    100644
        size    8193
        parent  34
        links   1
        pflags  40800000004
Indirect blocks:
               0 L0 EMBEDDED et=0 2200L/22P B=3046 ZSTD:size=26:version=10405:level=10:EMBEDDED

                segment [0000000000000000, 0000000000002200) size 8.50K
```

For each record, if it is ZSTD compressed, the parameters of the decoded
compression header are printed

Sponsored-By: The FreeBSD Foundation
Signed-off-by: Allan Jude <allanjude@freebsd.org>

Crow-Control commented on cmd/zdb/zdb.c:
```diff
@@ -27,6 +27,10 @@
  * Copyright (c) 2017, 2018 Lawrence Livermore National Security, LLC.
  * Copyright (c) 2015, 2017, Intel Corporation.
  * Copyright (c) 2020 Datto Inc.
+ * Copyright (c) 2020 The FreeBSD Foundation
+ *
+ * Portions of this software were developed by Allan Jude
+ * under sponsorship from the FreeBSD Foundation.
```
You sure legal copyright ownership lays with the FreeBSD foundation for your work? You have a transfer of ownership contract (as part of your sponsorship)?
If not, you are personal the legal copyright holder, not the foundation.

I doubt the headers are meant for everyone to (byond copyright) add they are the one working for their boss in an added note though.

allanjude replied:
The copyright is assigned as part of the contract, this specific text is also a requirement of the contract.

Crow-Control replied:
@allanjude Loud and clear, thanks for clearifying.

c0d3z3r0 replied:
If there aren't any arguments against, I'd update this to " * Copyright (c) 2020 The FreeBSD Foundation. All rights reserved." to keep it consistent

Crow-Control replied:
Thats legally no issue at all, as discussed before... :)

allanjude replied:
The FreeBSD Project, and FreeBSD Foundation specifically do not use the text 'all rights reserved' because it is not meaningful anymore

see: https://svnweb.freebsd.org/base?view=revision&revision=333391

c0d3z3r0 replied:
Oh, interesting! Would you suggest to adapt the others, too, then?

allanjude replied:
> 
> 
> Oh, interesting! Would you suggest to adapt the others, too, then?

We cannot change the copyright messages owned by others, but you can update your own to remove the problematic clause.

Crow-Control replied:
> We cannot change the copyright messages owned by others, but you can update your own to remove the problematic clause.

@allanJude the discussion for that has been finished already and i'm not interested in reopening it after months. It's misinformation to suggest one can't change copyright headers for someone else. 
(nor is it a "problematic" clause. it's just legally redundant)

Copyright headers are barely "meaningfull" to begin with, "All rights reserved" indeed doesn't have add additional meaning, either leaving it or removing it is irrelevant and perfectly fine.
(one could argue "all right reserved" is actually a false statement, because it isn't actually all rights reserved under US law, but licenced code under a specific licence)

They aren't needed to begin with, nor do they give any additional legal security. We can perfectly fine change them from everyone. Because there is no legal consequence nor legal precedent stating we can't.

As discussed before we can either remove them from all to standardise it (AGAIN) or leave it with everything (as done before). Luckily no one is legally able to enforce or require copyright header formatting on top of the licence anyway.

To be quite frank: I prefer not including "all rights reserved" in copyright headers myself (because it doesn't add anything legally speaking)

Anyhow, I think this is not the place to discuss it anyhow.

*edit*
Heavily restructured and rephrased.

Ornias1993 replied:
Copyright headers are barely "meaningfull" to begin with, "All rights reserved" doesn't have a meaning, either leaving it or removing it is irrelevant and perfectly fine.

BrainSlayer commented on cmd/zdb/zdb.c:
```diff
@@ -1557,6 +1562,64 @@ blkid2offset(const dnode_phys_t *dnp, const blkptr_t *bp,
+		decode_embedded_bp_compressed(bp, buf);
+		zstd_hdr = (zfs_zstdhdr_t *)buf;
+		zstd_hdr->c_len = BE_32(zstd_hdr->c_len);
+		zstd_hdr->raw_version_level = BE_32(zstd_hdr->raw_version_level);
```
i hope this buffer is not reused a second time. you are changing the content here while doing endian conversation

c0d3z3r0 replied:
what's the problem? buf is a local buffer and is a copy

allanjude replied:
I might be worth adding a comment here, as having done the byte swap, if you actually feed this to zio_decompress it will byte swap it again, and bad things will happen.

BrainSlayer replied:
@c0d3z3r0 my fault

allanjude replied:
I have a fix pushed already, it was a good catch

BrainSlayer replied:
oh okay. then it not my fault :-)

BrainSlayer commented on cmd/zdb/zdb.c:
```diff
@@ -1557,6 +1562,64 @@ blkid2offset(const dnode_phys_t *dnp, const blkptr_t *bp,
+	buf = abd_borrow_buf_copy(pabd, BP_GET_LSIZE(bp));
+	zstd_hdr = (zfs_zstdhdr_t *)buf;
+	zstd_hdr->c_len = BE_32(zstd_hdr->c_len);
+	zstd_hdr->raw_version_level = BE_32(zstd_hdr->raw_version_level);
```
same as above

c0d3z3r0 commented on include/sys/zstd/zstd.h:
```diff
@@ -34,6 +34,30 @@
+	/*
+	 * Version and compression level
+	 * We have to choose a union here to handle endian conversation
+	 * correctly, since the version and level is bitmask encoded.
```
Note to myself: /level is bitmask/ s/is/are/

allanjude committed:
Remove "all rights reserved" from old copyrights that I control

Signed-off-by: Allan Jude <allanjude@freebsd.org>

allanjude committed:
Improve comment about zstd compression header

Signed-off-by: Allan Jude <allanjude@freebsd.org>

c0d3z3r0 commented on cmd/zdb/zdb.c:
```diff
@@ -27,7 +27,7 @@
  * Copyright (c) 2015, 2017, Intel Corporation.
  * Copyright (c) 2020 Datto Inc.
- * Copyright (c) 2020 The FreeBSD Foundation
+ * Copyright (c) 2020 The FreeBSD Foundation.
```
let's add a * or [1] here, referring to the note below

allanjude committed:
Add a footnote to the FreeBSD Foundation copyright

Connect it to the statement about sponsored work

Signed-off-by: Allan Jude <allanjude@freebsd.org>

BrainSlayer commented on cmd/zdb/zdb.c:
```diff
@@ -1557,6 +1562,64 @@ blkid2offset(const dnode_phys_t *dnp, const blkptr_t *bp,
+		(void) snprintf(blkbuf + strlen(blkbuf),
+		    buflen - strlen(blkbuf),
+		    " ZSTD:size=%u:version=%u:level=%u:EMBEDDED",
+		    zstd_hdr.c_len, zstd_hdr.version, zstd_hdr.level);
```
buf is allocated. then copies to a zstd_hdr. then you return. hint: memleak

allanjude replied:
zstd_hdr is a stack variable, it goes out of scope once we return from this function. Notice it is not a pointer anymore.

BrainSlayer replied:
i'm talking about "buf" and not the stack variable. buf is allocated but never free'd

allanjude replied:
Ahh yes, fixed.

allanjude committed:
Don't leak memory when decoding an ZSTD compressed embedded BP

Signed-off-by: Allan Jude <allanjude@freebsd.org>

allanjude committed:
Add kstat counters for ZSTD

This will allow us to measure the number of times compression is
skipped due to a memory allocation failure, as well as the number of
times that we have used the fallback decompression context.

It also counts all other failure types:
- (de)compression level invalid
- (de)compression memory allocation failed
- (de)compression failed (zstd error)
- Failed to decode zstd_header for decompression (invalid size)

Sponsored By: The FreeBSD Foundation
Signed-off-by: Allan Jude <allanjude@freebsd.org>

Crow-Control commented on include/sys/zstd/zstd.h:
```diff
@@ -25,6 +25,10 @@
  * Copyright (c) 2016-2018, Allan Jude.
  * Copyright (c) 2018-2019, Sebastian Gottschall. All rights reserved.
  * Copyright (c) 2019-2020, Michael Niewöhner. All rights reserved.
+ * Copyright (c) 2020 The FreeBSD Foundation.
+ *
+ * Portions of this software were developed by Allan Jude
+ * under sponsorship from the FreeBSD Foundation.
```
Same footer formatting might be nice here...

c0d3z3r0 replied:
well need another run of copyright formatting fixup anyways
